File: code/SWE2-Project/Animal Tracker/src/main/java/edu/westga/cs3211/animaltracker/model/Sighting.java
package edu.westga.cs3211.animaltracker.model;

import java.time.LocalDateTime;

/**
 * Represents a documented sighting of an animal.
 *
 * @author mrocker1
 */
public class Sighting {
    private final int animalTagID;
    private final String location;
    private final double latitude;
    private final double longitude;
    private final LocalDateTime time;
    private final String notes;
    private final String username;

    /**
     * Instantiates a new Sighting object.
     *
     * @param animalID  the animal ID (mandatory)
     * @param location  the general location of the sighting (mandatory)
     * @param latitude  the latitude coordinate (mandatory)
     * @param longitude the longitude coordinate (mandatory)
     * @param time      the specific date and time of the sighting (mandatory)
     * @param notes     additional notes regarding the sighting (mandatory)
     * @param username  the username (optional), The reason it is optional is that
     *                  when creating the sighting locally, we do not store the
     *                  username and do not want to request the server to get it.
     */
    public Sighting(int animalID, String location, double latitude, double longitude, LocalDateTime time, String notes, String username) {
        if (animalID < 0) {
            throw new IllegalArgumentException("You must have an animal ID greater than zero");
        }
        if (isInvalidLocation(location)) {
            throw new IllegalArgumentException("Location must be provided.");
        }

        if (isInvalidLatitude(latitude)) {
            throw new IllegalArgumentException("Latitude must be a valid coordinate between -90 and 90.");
        }
        if (isInvalidLongitude(longitude)) {
            throw new IllegalArgumentException("Longitude must be a valid coordinate between -180 and 180.");
        }

        if (isInvalidTime(time)) {
            throw new IllegalArgumentException("Sighting time cannot be in the future.");
        }

        if (isInvalidNotes(notes)) {
            throw new IllegalArgumentException("Notes must contain valid text if provided.");
        }

        if (time == null) {
            throw new IllegalArgumentException("Sighting time cannot be null.");
        }

        if (notes == null) {
            throw new IllegalArgumentException("Notes cannot be null");
        }

        if (notes.isBlank()) {
            throw new IllegalArgumentException("Notes must contain valid text.");
        }

        this.animalTagID = animalID;
        this.location = location;
        this.latitude = latitude;
        this.longitude = longitude;
        this.time = time;
        this.notes = notes;
        this.username = username;
    }

    private static boolean isInvalidLocation(String location) {
        return location == null || location.trim().isEmpty();
    }

    private static boolean isInvalidNotes(String notes) {
        return notes != null && !notes.isEmpty() && notes.trim().isEmpty();
    }

    private static boolean isInvalidTime(LocalDateTime time) {
        return time != null && time.isAfter(LocalDateTime.now());
    }

    private static boolean isInvalidLongitude(double longitude) {
        return longitude < -180.0 || longitude > 180.0;
    }

    private static boolean isInvalidLatitude(double latitude) {
        return latitude < -90.0 || latitude > 90.0;
    }

    /**
     * Gets the animal.
     * @return the animal
     */
    public int getAnimalTagID() {
        return this.animalTagID;
    }

    /**
     * Gets the location.
     * @return the location
     */
    public String getLocation() {
        return this.location;
    }

    /**
     * Gets the latitude coordinate.
     * @return the latitude
     */
    public double getLatitude() {
        return this.latitude;
    }

    /**
     * Gets the longitude coordinate.
     * @return the longitude
     */
    public double getLongitude() {
        return this.longitude;
    }

    /**
     * Gets the time of the sighting.
     * @return the time
     */
    public LocalDateTime getTime() {
        return this.time;
    }

    /**
     * Gets the notes for the sighting.
     * @return the notes
     */
    public String getNotes() {
        return this.notes;
    }

    /**
     * Gets the Username.
     * @return Username
     */
    public String getUsername() {
        return this.username;
    }

    @Override
    public String toString() {
        return String.format("Sighting: %s at %s (%.5f, %.5f) with Notes: %s by %s at %s", this.animalTagID, this.location, this.latitude, this.longitude, this.notes, this.username, this.time);
    }
}

File: code/SWE2-Project/Animal Tracker Server/src/model/Sighting.py
from typing import Optional


class Sighting:
    """
    Represents a documented sighting of an animal.
    """

    def __init__(self, animal_tag, username, location, latitude, longitude, time, notes):
        """
        Initializes an instance of a Sighting.
        :param animal_tag: Tag for an animal.
        :param location: Location of sighting.
        :param latitude: Latitude of sighting.
        :param longitude: Longitude of sighting.
        :param time: Time of sighting.
        :param notes: Notes of sighting.
        """
        if animal_tag is None:
            raise ValueError("animal_tag must be provided.")

        if username is None:
            raise ValueError("username must be provided.")

        if self._is_invalid_location(location):
            raise ValueError("Location must be provided.")

        if self._is_invalid_latitude(latitude):
            raise ValueError("Latitude must be a valid coordinate between -90 and 90.")

        if self._is_invalid_longitude(longitude):
            raise ValueError("Longitude must be a valid coordinate between -180 and 180.")

        if self._is_invalid_time(time):
            raise ValueError("Sighting time must be provided.")

        if self._is_invalid_notes(notes):
            raise ValueError("Notes must contain valid text.")

        self._animal_tag = animal_tag
        self._username = username
        self._location = location
        self._latitude = latitude
        self._longitude = longitude
        self._time = time
        self._notes = notes

    @staticmethod
    def _is_invalid_location(location):
        """
        Checks if the location is invalid.
        :param location: Location of sighting.
        :return: True or False on whether the location is invalid.
        """
        return location is None or location.strip() == ""

    @staticmethod
    def _is_invalid_notes(notes: Optional[str]) -> bool:
        """
        Checks if the notes are invalid.
        :param notes: Notes of animal sighting.
        :return: True or False on whether the notes are invalid.
        """
        return notes is None or notes.strip() == ""

    @staticmethod
    def _is_invalid_time(time):
        """
        Checks if the time is invalid.
        :param time: Time of animal sighting.
        :return: True or False on whether the time is invalid.
        """
        return time is None

    @staticmethod
    def _is_invalid_longitude(longitude):
        """
        Checks if the longitude is invalid.
        :param longitude: Longitude of animal sighting.
        :return: True or False on whether the longitude is invalid.
        """
        return longitude < -180.0 or longitude > 180.0

    @staticmethod
    def _is_invalid_latitude(latitude):
        """
        Checks if the latitude is invalid.
        :param latitude: Latitude of animal sighting.
        :return: True or False on whether the latitude is invalid.
        """
        return latitude < -90.0 or latitude > 90.0

    def get_username(self):
        return self._username

    def get_animal_tag(self):
        """
        Gets the animal tag.
        :return: The animal tag.
        """
        return self._animal_tag

    def get_location(self):
        """
        Gets the location.
        :return: The location of sighting.
        """
        return self._location

    def get_latitude(self):
        """
        Gets the latitude.
        :return: The latitude of sighting.
        """
        return self._latitude

    def get_longitude(self):
        """
        Gets the longitude.
        :return: The longitude of sighting.
        """
        return self._longitude

    def get_time(self):
        """
        Gets the time of sighting.
        :return: The time of sighting.
        """
        return self._time

    def get_notes(self):
        """
        Gets the notes of sighting.
        :return: The notes of sighting.
        """
        return self._notes
